crates/wasmtime/src/instance/allocator.rs

```diff
@@ -0,0 +1,183 @@
+use crate::trampoline::StoreInstanceHandle;
+use crate::{Extern, Instance, Module, Store, Trap};
+use anyhow::{anyhow, bail, Context, Error, Result};
+use std::mem;
+use wasmtime_environ::wasm::EntityIndex;
+use wasmtime_jit::CompiledModule;
+use wasmtime_runtime::{
+    Imports, InstanceAllocationRequest, InstantiationError, StackMapRegistry, VMContext,
+    VMExternRefActivationsTable, VMFunctionBody,
+};
+
+mod default;
+
+pub use self::default::DefaultInstanceAllocator;
+
+/// Represents an instance allocator.
+pub trait InstanceAllocator: Send + Sync {
+    /// Allocates an instance for the given module.
+    fn allocate(&self, store: &Store, module: &Module, imports: &[Extern]) -> Result<Instance>;
+}
```

Comment:
The public version of the trait doesn't have any deallocation hooks? I don't see any mention of this in the RFC either. How is the instance pool supposed to recycle an instance's memory?

Reply:
The RFC doesn't change the lifetime semantics of module instances, so this trait doesn't need a `deallocate` method as it is assumed a store owns the returned instance.  When the store gets dropped, all associated instances will be deallocated like they are today.

This differs from the runtime `InstanceAllocator` trait as there's no concept of store there and handles are what are being allocated and deallocated.

Reply:
I feel like I'm missing something or we are talking past each other, sorry: When a store is dropped, doesn't the instance pool need to know that each of the store's instance's memory ranges are available for reuse now?

Reply:
It will via `InstanceHandle::dealloc` which is currently calling the associated allocator to deallocate the instance.  The pool then will know what associated memories and tables to put back on the free list.

I'm trying to remove the `allocator` field of `InstanceHandle` though since I don't like storing this unsafe back pointer to the allocator.

Reply:
Okay and that `dealloc` call fowards to `wasmtime_runtime::InstanceAllocator::deallocate`, but that doesn't have a corresponding method in the public `wasmtime::InstanceAllocator` trait. Do we expect the instance pool to use the internal `wasmtime-runtime` APIs instead of the public `wasmtime` APIs? Do we expect any implementations of `InstanceAllocator` other than the default and the instance pool? Do we expect to eventually add a `deallocate` method to `wasmtime::InstanceAllocator` as well?

Reply:
What I'd like to do is remove the `InstanceAllocator` and `DefaultInstanceAllocator` types from the `wasmtime` crate in favor of `Config` taking a `Arc<dyn wasmtime_runtime::InstanceAllocator>` directly.  This would both simply things and help me remove this back pointer from `wasmtime_runtime::Instance` to the allocator that created the instance so that deallocation can be handled directly from the drop implementation of `Store`.

I would then like to open up the construction of `wasmtime_runtime::Instance` and `wasmtime_runtime::InstanceHandle` such that instance allocators could be implemented in crates other than `wasmtime_runtime`.

However, given our stance on `wasmtime` being the crate we offer semantic versioning and `wasmtime_runtime` being an implementation details crate, I don't know if this is feasible to do.

Reply:
I've went ahead with an enum-based solution to the problem as discussed on the RFC (which I'll be amending shortly).

As such, the `InstanceAllocator` trait has been removed from the `wasmtime` crate.